File: src/mcs/lib/Shared/Test/Test_Bug.cs
//----------------------------------------------------------------------------
//
//       Copyright (c) 2019 Frank Eskesen.
//
//       This file is free content, distributed under the Lesser GNU
//       General Public License, version 3.0.
//       (See accompanying file LICENSE.LGPL-3.0 or the original
//       contained within https://www.gnu.org/licenses/lgpl-3.0.en.html)
//
//----------------------------------------------------------------------------
//
// Title-
//       Test_Bug.cs
//
// Purpose-
//       Test library Debug object.
//
// Last change date-
//       2019/02/15
//
//----------------------------------------------------------------------------
using System;
using System.Diagnostics;           // For demonstration purposes only.
using Shared;                       // For Debug

//============================================================================
/// We are not reall using System.Diagnostics, it is included only to
/// demonstrate how to use this class when it is included.
//============================================================================
using Debug = Shared.Debug;

//============================================================================
// Test cases
//============================================================================
static class Test {
//----------------------------------------------------------------------------
//
// Method-
//       test_debug
//
// Purpose-
//       Debug.cs minimal verification test.
//
//----------------------------------------------------------------------------
public static void test_debug( ) { // Test Debug.cs
   //-------------------------------------------------------------------------
   using( Debug me= new Debug("debug.log") ) {
       me.putLine("Debug: using( Debug x= new Debug({0}) {1}",
                  "\"debug.log\"", "{...}");
       Test.debug_unit();
   }

   GC.Collect();
   GC.WaitForPendingFinalizers();
   Debug.assert( Debug.debug == null );

   //-------------------------------------------------------------------------
   using( Debug me= new Debug() ) {
       me.putLine("\nDebug: Functional tests");
       me.putLine();
       me.putLine("Tested: empty putLine");

       Test.debug_null();
   }

   //-------------------------------------------------------------------------
   {
       Debug me= new Debug();
       me.putLine("\nDebug: Debug x= new Debug(): {0} Debug.close()", "{...}");
       Test.debug_unit();
       me.close();
       me = null;
   }

   GC.Collect();
   GC.WaitForPendingFinalizers();
   Debug.assert( Debug.debug == null );

#if false // THIS TEST FAILS // TODO: Investigate.
   //-------------------------------------------------------------------------
   Debug.WriteLine("\nDebug: Debug x= new Debug(): {0} NO close()", "{...}");
   {
       Debug me= new Debug();
       Test.debug_unit();
   }

   GC.Collect();
   GC.WaitForPendingFinalizers();
   Debug.assert( Debug.debug == null );
#endif
}

static void debug_unit( ) {         // Run Debug unit test
   Debug.debug.put("debug_unit: First test line...\n");

   Debug.debug.put("Begin a console+log line, ");
   Debug.debug.putLine("end the console+log line.");

   Debug.debug.log("Begin a log-only line, ");
   Debug.debug.logLine("end the log-only line.");

   Debug.DEBUGGING= false;
   Debug.debug.put("Begin a discarded line, ");
   Debug.debug.putLine("end the discarded line.");

   Debug.Write("Begin an unconditional console+log line, ");
   Debug.WriteLine("end the unconditional line.");

   Debug.DEBUGGING= true;
   Debug.debug.put("debug_unit: ...Last test line\n");
}

static void debug_null( ) {         // Run null object test
   object null_object= null;

   // Missing references compile but throw a run-time exception
// Debug.debug.put("Testing missing({0}) reference\n");

   // null references compile but throw a run-time exception
// Debug.debug.put("Testing missing({0}) reference\n", null);

   // Null objects compile are display the same as an empty string
   Debug.debug.put("Testing null({0}) reference\n", null_object);
}
}  // static class Test

//============================================================================
// PROGRAM: Mainline code
//============================================================================
public class Program {
static void Main(string[] args)     // Mainline code
{
   Test.test_debug();               // Run all tests
}
}   // class Program

File: src/cpp/lib/pub/Test/Test_bug.cpp
//----------------------------------------------------------------------------
//
//       Copyright (c) 2007-2026 Frank Eskesen.
//
//       This file is free content, distributed under the GNU General
//       Public License, version 3.0.
//       (See accompanying file LICENSE.GPL-3.0 or the original
//       contained within https://www.gnu.org/licenses/gpl-3.0.en.html)
//
// SPDX-License-Identifier: GPL-3.0-only
//----------------------------------------------------------------------------
//
// Title-
//       Test_bug.cpp
//
// Purpose-
//       Test debugging methods.
//
// Last change date-
//       2026/02/23
//
//----------------------------------------------------------------------------
#include <stdexcept>                // For std::runtime_error
#include <cerrno>                   // For errno

#include <pub/TEST.H>               // For test functions and macros
#include "pub/Debug.h"              // For Debug, tested
#include <pub/Wrapper.h>            // For class Wrapper

#define PUB _LIBPUB_NAMESPACE
using PUB::Debug;
using namespace PUB::debugging;
using PUB::Wrapper;                 // For pub::Wrapper class

//----------------------------------------------------------------------------
// Constants for parameterization
//----------------------------------------------------------------------------
enum
{  HCDM= false                      // Hard Core Debug Mode?
,  VERBOSE= 0                       // Verbosity, higher is more verbose
}; // enum

//----------------------------------------------------------------------------
// Extended options
//----------------------------------------------------------------------------
//static struct option   opts[]=      // The getopt_long parameter: longopts
//{  {0, 0, 0, 0}                     // (End of option list)
//};

//----------------------------------------------------------------------------
//
// Subroutine-
//       test_bt
//
// Purpose-
//       Test backtrace
//
//----------------------------------------------------------------------------
_LIBPUB_NOINLINE
static void bar() {
   debugf("Backtrace test\n");
   debug_backtrace();
   debugf("\n");
}

_LIBPUB_NOINLINE
static void foo() {
   bar();
}

_LIBPUB_NOINLINE
static void the() {
   foo();
}

_LIBPUB_NOINLINE
static void test_bt() {
   the();
}

//----------------------------------------------------------------------------
//
// Subroutine-
//       main
//
// Purpose-
//       Mainline code.
//
//----------------------------------------------------------------------------
extern int                          // Return code
   main(                            // Mainline code
     int               argc,        // Argument count (UNUSED)
     char*             argv[])      // Argument array (UNUSED)
{
   //-------------------------------------------------------------------------
   // Initialize
   Wrapper  tc;                     // The test case wrapper
   Wrapper* tr= &tc;                // A test case wrapper pointer

   tc.on_main([tr](int, char*[])
   {
     if( opt_verbose )
       debugf("%s: %s %s\n", __FILE__, __DATE__, __TIME__);

     int error_count= 0;

     // Test backtrace
     // Implementation note: backtrace is provided by the boost library and
     // its output varies depending upon the installed version. Since
     // regression testing checks our output, it's not tested by default.
     if( opt_verbose )
       test_bt();

     // Test modes
     debug_set_head(Debug::HEAD_TIME);
     debug_set_mode(Debug::MODE_DEFAULT);
     debugf("Standard mode:\n");
     debugf("This appears in %s and %s\n", "TRACE", "STDOUT");
     errorf("This appears in %s and %s\n", "TRACE", "STDERR");
     tracef("This appears in %s ONLY\n",   "TRACE");
     debugh("This appears in %s and %s\n", "TRACE", "STDOUT");
     errorh("This appears in %s and %s\n", "TRACE", "STDERR");
     traceh("This appears in %s ONLY\n",   "TRACE");

     debug_set_mode(Debug::MODE_IGNORE);
     errno= 0;
     debugf("Ignore mode:\n");
     errorf("Ignore mode:\n");
     tracef("Ignore mode:\n");
     debugh("Ignore mode:\n");
     errorh("Ignore mode:\n");
     traceh("Ignore mode:\n");

     debug_set_mode(Debug::MODE_INTENSIVE);
     debugf("Intensive mode:\n");
     debugf("This appears in %s and %s\n", "TRACE", "STDOUT");
     errorf("This appears in %s and %s\n", "TRACE", "STDERR");
     tracef("This appears in %s ONLY\n",   "TRACE");
     debugh("This appears in %s and %s\n", "TRACE", "STDOUT");
     errorh("This appears in %s and %s\n", "TRACE", "STDERR");
     traceh("This appears in %s ONLY\n",   "TRACE");

     // Test throwf (output not checked)
     if( opt_verbose ) {
       bool caught= false;
       try {
         debugf("\n%4d Testing throwf\n", __LINE__);
         throwf("%4d runtime_error", __LINE__);
         error_count += MUST_NOT(fail to throw an exception);
       } catch(std::runtime_error& X) {
         debugf("(Backtrace expected)\n");
         debugf("%4d ..As expected: %s\n", __LINE__, X.what());
         caught= true;
       }
       error_count += VERIFY( caught == true );
     }

     // Test abortf (output not checked)
     if( opt_verbose > 1 ) {
       try {
         debugf("\n%4d Testing abortf\n", __LINE__);
         abortf("%4d abortf test", __LINE__);
         error_count += MUST_NOT(fail to abort);
       } catch(...) {
         error_count += MUST_NOT(fail to abort);
       }
       error_count += MUST_NOT(fail to abort);
     }

     return error_count != 0;
   });

   //-------------------------------------------------------------------------
   // Run the test
   setlocale(LC_NUMERIC, "");       // Activates ' thousand separator
   opt_hcdm= HCDM;
   opt_verbose= VERBOSE;

   return tc.run(argc, argv);
}
